File: unmastered/ctci_problems/linked_lists/2.1.js
var LinkedList = require('../../../mastered/data_structures/linked_list').LinkedList;

function removeDups(list){
	let seen = {};
	let trailingPtr = list.head;
	let ptr = trailingPtr.next;
	while (ptr !== list.tail){
		if (seen[ptr.val]){
			trailingPtr.next = ptr.next;
			ptr.next.prev = trailingPtr;
		} else {
			seen[ptr.val] = true;
			trailingPtr = ptr;
		}
		ptr = ptr.next;
	}
}

function removeDupsConstantSpaceWithOrderChange(list){
	list.quickSort();
	let ptr = list.head.next;
	while(ptr !== list.tail){
		let prev = ptr.prev;
		let next = ptr.next
		if (prev.val === ptr.val){
			prev.next = next;
			next.prev = prev;
		}
		ptr = next;
	}
}

function removeDupsConstantSpaceNoOrderChange(list){
	let trailingPtr = list.head.next;
	let leadingPtr;
	while (trailingPtr !== list.tail){
		leadingPtr = trailingPtr.next;
		while (leadingPtr !== list.tail){
			let nextRound = leadingPtr.next
			if (trailingPtr.val === leadingPtr.val){				
				list.deleteNode(leadingPtr);
			} 
			leadingPtr = nextRound;
		}
		trailingPtr = trailingPtr.next;
	}
}


//TEST

// function initializeTest(){
// 	let test = new LinkedList();
// 	test.addNode(3);
// 	// test.addNode(5);
// 	test.addNode(3);
// 	// test.addNode(-3);
// 	// test.addNode(10);
// 	// test.addNode(8);
// 	return test;
// }

// let test1 = initializeTest();
// test1.print();
// console.log("before")
// removeDups(test1);
// test1.print();
// console.log("after")

// console.log("next test --------")

// let test2 = initializeTest();
// test2.print();
// console.log("before")
// removeDupsConstantSpaceNoOrderChange(test2);
// test2.print();
// console.log("after")

File: unmastered/ctci_problems/linked_lists/2.1.py
def removeDups(head):
	currOriginal = head
	while ((not currOriginal is None) and (not currOriginal.next is None)):
		parent = currOriginal;
		currNode = parent.next;
		while (not currNode is None):
			if (currNode.val == currOriginal.val):
				parent.next = currNode.next;
			else:
				parent = currNode;

			currNode = currNode.next

		currOriginal = currOriginal.next;



class Node(object):
	def __init__(self, val):
		self.val = val;
		self.next = None


def printList(head):
	currNode = head;
	while (not currNode is None):
		print currNode.val;
		currNode = currNode.next


def test():
	head = Node(1);
	head.next = Node(1);
	head.next.next = Node(2);
	head.next.next.next = Node(3);
	head.next.next.next.next = Node(2);
	removeDups(head);
	printList(head);
	#expect 123

#test()


# def faketest():
# 	x = Node(1);
# 	y = Node(2);
# 	x.next = y;
# 	print (x.next == y)

# faketest()